### PR TITLE
(PUP-8942) Support issuing certificates with IP SAN

### PR DESCRIPTION
**Pull request by lazartea**
Currently, there is support for signing CSRs with a DNS subject alternative name, but there is no option to generate a cert with an IP SAN. We are also currently using the ``--dns-alt-names`` flag, rather than using ``--subject-alt-names`` and having the user specify the altname. This ticket adds IP SAN functionality, and aliases -dns-alt-names to ``--subject-alt-names``. Users who do not specify their altname type will default to DNS, in order to support the old format.

### CODE REVIEW

File: lib/puppet/ssl/certificate_authority.rb [hunk 1]
```diff
@@ -121,7 +121,11 @@ def generate(name, options = {})
     raise ArgumentError, _("A Certificate already exists for %{name}") % { name: name } if Puppet::SSL::Certificate.indirection.find(name)
 
     # Pass on any requested subjectAltName field.
-    san = options[:dns_alt_names]
+    if options[:dns_alt_names] && options[:subject_alt_names]
+      san = options[:dns_alt_names] + ", " + options[:subject_alt_names]
+    else
+      san = options[:dns_alt_names] || options[:subject_alt_names]
+    end
```
Comment (Magisus) [inline lib/puppet/ssl/certificate_authority.rb]: @ahpook do we want to deprecate (i.e. actually issue a deprecation warning) the old flag? Or do we intend to maintain both going forward?

File: spec/unit/ssl/certificate_request_spec.rb [hunk 2]
```diff
@@ -177,6 +177,17 @@
       end
     end
 
+    context "with DNS and IP SAN specified" do
+      before :each do
+        Puppet[:dns_alt_names] = ""
```
Comment (Magisus) [inline spec/unit/ssl/certificate_request_spec.rb]: Do we want to take this opportunity to rename/alias the setting as well?

File: spec/unit/ssl/certificate_request_spec.rb [hunk 3]
```diff
@@ -177,6 +177,17 @@
...
+      end
+
+      it "should return the subjectAltName values" do
+        request.generate(key, :dns_alt_names => 'DNS:foo, bar, IP:172.16.254.1')
```
Comment (Magisus) [inline spec/unit/ssl/certificate_request_spec.rb]: Similar question here, should we rename this option to `:subject_alt_names` as well? (this one is only internal, not user-facing)

File: spec/unit/ssl/certificate_request_spec.rb (hunk 2 above)
Comment (Magisus) [reply spec/unit/ssl/certificate_request_spec.rb]: Given the use case described in the docs here https://puppet.com/docs/puppetserver/5.3/scaling_puppet_server.html#creating-and-configuring-compile-masters I'm inclined to say yes.

Comment (Magisus) [reply spec/unit/ssl/certificate_request_spec.rb]: The definition of the `dns_alt_names` setting can be found here https://github.com/puppetlabs/puppet/blob/master/lib/puppet/defaults.rb#L745. There's a big doc there that at minimum should be updated to describe the new tagged syntax. I think it would be good to also deprecate that setting in favor of one called `subject_alt_names`. For an example of a deprecated setting, see https://github.com/puppetlabs/puppet/blob/master/lib/puppet/defaults.rb#L1814-L1823.

**Comment by lazartea** [comment]
Should we be renaming/aliasing the ``--allow-dns-alt-names`` flag too?

File: lib/puppet/application/cert.rb [hunk 4]
```diff
@@ -302,6 +302,8 @@ def setup
     if subcommand == :generate
       if Puppet.settings.set_by_cli?(:dns_alt_names)
         options[:dns_alt_names] = Puppet[:dns_alt_names]
+      elsif Puppet.settings.set_by_cli?(:subject_alt_names)
+        options[:subject_alt_names] = Puppet[:subject_alt_names]
```
[the remark this answers is not in the document]
Comment (Magisus) [reply lib/puppet/application/cert.rb]: What kind of error?

[the remark this answers is not in the document]
Comment (lazartea) [reply lib/puppet/application/cert.rb]: It's says that subject-alt-names requires a parameter, which makes me think that it's feeding the comma-separated list through to dns-alt-names first and then subject-alt-names seems like it doesn't have any params

**Comment by Magisus** [comment]
I'd be in favor of also updating the `--allow` flag.

File: lib/puppet/ssl/certificate_request.rb [hunk 5]
```diff
@@ -267,7 +267,17 @@ def extension_request_attribute(options)
 
     if options[:dns_alt_names]
       names = options[:dns_alt_names].split(/\s*,\s*/).map(&:strip) + [name]
-      names = names.sort.uniq.map {|name| "DNS:#{name}" }.join(", ")
+
+      names = names.sort.map do |name|
+        if (!name.start_with? "IP:") && (!name.start_with? "DNS:")
```
Comment (justinstoller) [inline lib/puppet/ssl/certificate_request.rb]: You don't need parens around each clause, though you should put them around each arg to `start_with?`.

File: lib/puppet/ssl/certificate_request.rb [hunk 6]
```diff
@@ -267,7 +267,17 @@ def extension_request_attribute(options)
...
+          "DNS:#{name}"
+        else
+          "#{name}"
```
Comment (justinstoller) [inline lib/puppet/ssl/certificate_request.rb]: You don't need to interpolate this. It's equivalent to `name.to_s` and if `name` isn't already a String `start_with?` above will raise.

File: lib/puppet/ssl/certificate_request.rb [hunk 7]
```diff
@@ -267,7 +267,17 @@ def extension_request_attribute(options)
...
+        end
+      end
+
+      names = names.uniq.join(", ")
```
Comment (justinstoller) [inline lib/puppet/ssl/certificate_request.rb]: We shouldn't assign the same variable too many times in a method. We might want to combine this and the above statement and name the first assignment `raw_names` and the second `parsed_names` or something.

File: lib/puppet/defaults.rb [hunk 8]
```diff
@@ -742,10 +742,10 @@ def self.default_basemodulepath
...
         Defaults to the node's fully qualified domain name.",
       :hook => proc { |value| raise(ArgumentError, _("Certificate names must be lower case")) unless value == value.downcase }},
-    :dns_alt_names => {
-      :default => '',
-      :desc    => <<EOT,
-A comma-separated list of alternate DNS names for Puppet Server. These are extra
+    :subject_alt_names => {
+        :default => '',
+        :desc => <<EOT,
+A comma-separated list of alternate names for Puppet Server. These are extra
```
Comment (justinstoller) [inline lib/puppet/defaults.rb]: If this is now for `subject_alt_names` we should update this to mention that it can be `IP` or `DNS` type when prefixed with the type and a colon, with un-typed inputs defaulting to the `DNS` type.

File: lib/puppet/ssl/certificate_request.rb (hunk 7 above)
Comment (lazartea) [reply lib/puppet/ssl/certificate_request.rb]: The reason I split it out into two statements was because if we sort and append the DNS prefix in the same statement, then we see the duplication issue @Magisus mentioned. Or did you mean something else? Either way, I'll fix the naming issue.

Comment (justinstoller) [reply lib/puppet/ssl/certificate_request.rb]: No meant combining them, I remember there be a duplication issue, but I can't imagine how chaining those methods would cause it:
```rb
irb(main):011:0> foo = ['baba', 'titi', 'dodo', 'baba', 'papa', 'mama', 'gigi', 'mama']
=> ["baba", "titi", "dodo", "baba", "papa", "mama", "gigi", "mama"]
irb(main):012:0> foo.sort.map{|n| "PREFIX:#{n}" }.uniq.join(", ")
=> "PREFIX:baba, PREFIX:dodo, PREFIX:gigi, PREFIX:mama, PREFIX:papa, PREFIX:titi"
```

File: lib/puppet/ssl/certificate_request.rb [hunk 9]
```diff
@@ -267,7 +267,17 @@ def extension_request_attribute(options)
 
     if options[:dns_alt_names]
       names = options[:dns_alt_names].split(/\s*,\s*/).map(&:strip) + [name]
-      names = names.sort.uniq.map {|name| "DNS:#{name}" }.join(", ")
+
+      names = names.sort.map do |name|
```
Comment (justinstoller) [inline lib/puppet/ssl/certificate_request.rb]: It's a good thing to sort, but with the inputs now being (potentially) prefixed it might make sense to sort after we ensure they're prefixed.

File: lib/puppet/ssl/certificate_request.rb (hunk 7 above)
Comment (lazartea) [reply lib/puppet/ssl/certificate_request.rb]: Ah, I think the problem was that originally we were calling .uniq before we added the prefix. I'll just move it to afterwards like you did.

File: lib/puppet/application/cert.rb [hunk 10]
```diff
@@ -71,7 +71,12 @@ def subcommand=(name)
   end
 
   option("--[no-]allow-dns-alt-names") do |value|
-    options[:allow_dns_alt_names] = value
+    options[:allow_subject_alt_names] = value
```
Comment (Magisus) [inline lib/puppet/application/cert.rb]: Need to be sure to update the places where this is pulled out of the options hash to use the new name.

File: lib/puppet/face/certificate.rb [hunk 11]
```diff
@@ -113,27 +113,27 @@
       $ puppet certificate sign somenode.puppetlabs.lan --ca-location remote
     EOT
 
-    option("--[no-]allow-dns-alt-names") do
+    option("--[no-]allow-subject-alt-names") do
```
Comment (Magisus) [inline lib/puppet/face/certificate.rb]: I think we will need the same aliasing and deprecating here as we have in the other command.

Comment (lazartea) [reply lib/puppet/face/certificate.rb]: I just added the deprecation warning, but I'm not sure if/how we need to alias allow-dns-alt-names? It looks like we're  using ``options[:allow_subject_alt_names]`` which should have already been aliased if they entered --allow-dns-alt-names, right?

**Comment by lazartea** [comment]
I've manually tested that the deprecation warning comes up, but I haven't managed to actually check if the cert is being signed yet. I'm looking into disabling autosigning when I ``puppet cert generate`` with the``--subject_alt_names`` flag. I've edited my ``puppet.conf`` file like mentioned [here](https://puppet.com/docs/puppet/5.3/ssl_autosign.html#disabling-autosigning) but that didn't seem to solve the problem. Any ideas?

**Comment by justinstoller** [comment]
`puppet cert generate` will generate a new certificate, not a csr (it will generate a csr in the process of generating a cert, but it will delete the csr after having signed the cert).

If you want to just create a pending csr to sign you might want to try running the agent with a new certname against a running master while autosigning is off.

File: lib/puppet/face/ca.rb [hunk 12]
```diff
@@ -168,8 +168,8 @@
 
   action :sign do
     summary _("Sign an outstanding certificate request.")
-    option("--[no-]allow-dns-alt-names") do
-      summary _("Whether or not to accept DNS alt names in the certificate request")
+    option("--[no-]allow-subject-alt-names") do
+      summary _("Whether or not to accept alt names in the certificate request")
```
Comment (Magisus) [inline lib/puppet/face/ca.rb]: This one should look like the one for the `cert` command up above: both options specified, pointing to the same hash key, `dns` with a deprecation warning.

**Comment by Magisus** [comment]
So generate creates an actual cert for the specified host, so it will create a CSR then sign it, always. To try out the new flag, you need to generate just a certificate request with alt names, which you can do by editing the `[agent]` section of `puppet.conf` to list some alt names via the setting, then sign it manually using `be puppet cert sign` with the new flag (or the old one to check the warning). For this you will need a server running to accept the request.

File: lib/puppet/face/certificate.rb [hunk 13]
```diff
@@ -114,7 +114,12 @@
     EOT
 
     option("--[no-]allow-subject-alt-names") do
-      summary _("Whether or not to accept DNS alt names in the certificate request")
+      summary _("Whether or not to accept subject alt names in the certificate request")
+    end
+
+    option("--[no-]allow-dns-alt-names") do
+      summary _("Whether or not to accept subject alt names in the certificate request")
+      Puppet.deprecation_warning(_("--allow-dns-alt-names is deprecated and has been replaced by --allow-subject-alt-names"))
```
Comment (lazartea) [inline lib/puppet/face/certificate.rb]: I've just noticed some weirdness with this. I did ``be puppet help`` and this deprecation warning came up... I'm not sure what's going on but I'm going to dig into it further.

**Comment by lazartea** [comment]
@Magisus @justinstoller thanks for the help! I've manually tested both flags now, and they both are working (and the old one has the deprecation flag).

File: lib/puppet/application/cert.rb (hunk 10 above)
Comment (Magisus) [reply lib/puppet/application/cert.rb]: Looks like this one got updated to "subject" by accident.

File: lib/puppet/face/certificate.rb [hunk 14]
```diff
@@ -53,9 +53,9 @@
 
     # Duplicate the option here explicitly to distinguish if it was passed arg
     # us vs. set in the config file.
-    option "--dns-alt-names "+ _("NAMES") do
-      summary _("Additional DNS names to add to the certificate request")
-      description Puppet.settings.setting(:dns_alt_names).desc
+    option "--subject-alt-names "+ _("NAMES") do
+      summary _("Additional subject alt names to add to the certificate request")
+      description Puppet.settings.setting(:subject_alt_names).desc
     end
```
Comment (Magisus) [inline lib/puppet/face/certificate.rb]: I think we may need to keep `--dns-alt-names` here in addition (with a deprecation warning) like we are doing elsewhere, to avoid breaking people.

File: lib/puppet/face/certificate.rb [hunk 15]
```diff
@@ -69,9 +69,9 @@
       # command line. If both specified, we'll fail.
       # jeffweiss 17 april 2012
 
-      global_setting_from_cli = Puppet.settings.set_by_cli?(:dns_alt_names) == true
-      raise ArgumentError, _("Can't specify both --dns_alt_names and --dns-alt-names") if options[:dns_alt_names] and global_setting_from_cli
-      options[:dns_alt_names] = Puppet[:dns_alt_names] if global_setting_from_cli
+      global_setting_from_cli = Puppet.settings.set_by_cli?(:subject_alt_names) == true
+      raise ArgumentError, _("Can't specify both --subject_alt_names and --subject-alt-names") if options[:subject_alt_names] and global_setting_from_cli
+      options[:subject_alt_names] = Puppet[:subject_alt_names] if global_setting_from_cli
```
Comment (Magisus) [inline lib/puppet/face/certificate.rb]: It's unclear if this bit here is necessary to keep around (and the above comment doesn't make it any better). We should think through what's actually going on here and clean this up (including making the comment better if we can).

**Comment by Magisus** [comment]
jenkins please test this

File: lib/puppet/application/cert.rb (hunk 10 above)
Comment (lazartea) [reply lib/puppet/application/cert.rb]: I thought we wanted this one to be subject so that they pointed to the same hash?

**Comment by lazartea** [comment]
Five tests are currently failing due to a bug related to faces. Back in 2012, (#13898 for anyone looking), some logic was added to fix a bug where ``--dns_alt_names`` and ``--dns-alt-names`` were used at the same time with the certificate command. ``--dns_alt_names`` corresponded to the ``Puppet[:dns_alt_names]`` global setting, and ``--dns-alt-names`` corresponded to a local ``options[:dns_alt_names]``. If both were passed in, an error would be called.

However, adding in subject alt names would result in two more possible flags (``--subject_alt_names`` and ``--subject-alt-names``) which we would have to handle. This could lead to some pretty convoluted logic (I want to say there's something like 16 possible cases we'd have to account for?)

As this issue only affects faces, there are several ways we can go from here:

- implement IP support without changing the name of ``--dns-alt-names``. This is the easiest option, but does mean that we're using a flag that's kind of a misnomer. I'm going to put up a separate PR with this on its own. 
- change the name of ``--allow-subject-alt-names`` and ``--subject-alt-names`` only for cert. This would ignore all of the faces complications, but would require users to remember two separate flags based on context.
- change the name everywhere with deprecation warnings, leading to the continuation of the  gross logic mentioned above.
- some other magical solution that fixes everything

In any case, I think that this is clearly not a decision I should be making, so I'm marking this as blocked until we can get some more direction. @Magisus is there anything I've left out?

**Comment by Magisus** [comment]
That sounds right to me. @ahpook can you take a look at the situation Amy described and give us a feel for how to proceed? Since you're the most familiar with customers' preferred certificate workflows, you're the closest thing we have to UX person at this point.

**Comment by justinstoller** [comment]
What if we didn't make `subject_alt_names` a setting, but only added the cli flag `subject-alt-names`?

**Comment by Magisus** [comment]
Then it will only work for faces. All the flags for applications come from settings.

File: api/docs/http_certificate_status.md [hunk 16]
```diff
@@ -99,7 +99,7 @@ incomplete.
         "SHA256":"A6:44:08:A6:38:62:88:5B:32:97:20:49:8A:4A:4A:AD:65:C3:3E:A2:4C:30:72:73:02:C5:F3:D4:0E:B7:FC:2F",
         "SHA512":"CA:A0:8C:B9:FE:9D:C2:72:18:57:08:E9:4B:11:B7:BC:4E:F7:52:C8:9C:76:03:45:B4:B6:C5:D2:DC:E8:79:43:D7:71:1F:5C:97:FA:B2:F3:ED:AE:19:BD:A9:3B:DB:9F:A5:B4:8D:57:3F:40:34:29:50:AA:AA:0A:93:D8:D7:54"
       },
-      "dns_alt_names":["DNS:puppet","DNS:mycertname"]
+      "subject_alt_names":["DNS:puppet","DNS:mycertname"]
```
Comment (joshcooper) [inline api/docs/http_certificate_status.md]: Sorry, drive-by review. I don't think we should change the REST API schema in a .z release.

File: lib/puppet/defaults.rb [hunk 17]
```diff
@@ -742,14 +742,15 @@ def self.default_basemodulepath
 
         Defaults to the node's fully qualified domain name.",
       :hook => proc { |value| raise(ArgumentError, _("Certificate names must be lower case")) unless value == value.downcase }},
-    :dns_alt_names => {
-      :default => '',
+    :subject_alt_names => {
```
Comment (joshcooper) [inline lib/puppet/defaults.rb]: Unfortunately, we can't change setting names in a .z. Note each setting is automagically made available as a command line argument, so this could break anyone calling puppet programmatically with the old command line argument, which would now be unrecognized.

Comment (joshcooper) [reply lib/puppet/defaults.rb]: Nevermind, I see the `dns_alt_names` setting below.

File: api/docs/http_certificate_status.md (hunk 16 above)
Comment (Magisus) [reply api/docs/http_certificate_status.md]: I don't think this was actually in the scope of the ticket anyway. Looks like this may have been changed by accident, with an over-eager find/replace.

**Comment by Magisus** [comment]
Due to the removal of the Ruby CA in Puppet 6, any attempt to finish what this PR started (with renaming `dns_alt_names` to `subject_alt_names`) would need to be substantially reworked. Closing this.
